### ISSUE TITLE
Swagger UI unwrapping Spring Data `Pageable` with wrong field names

### ISSUE THREAD

**Issue by ArchangelX360** [springfox/springfox#2623]
My controller is the following:

```java
  @GetMapping
  public ResponseEntity getAll(final @PageableDefault(sort = "id") Pageable pageable) {
    return ResponseEntity.ok(null);
  }
```

The resulting Swagger UI is:
![screen shot 2018-08-15 at 17 54 22](https://user-images.githubusercontent.com/4418480/44158400-def5ac80-a0b4-11e8-9739-0ede8dc03943.png)

It looks like it uses the `Pageable` getters to infer those parameters, bypassing `springfox-data-rest` configuration.

I am using Spring Boot 1.5.10.RELEASE and Springfox 2.9.2 with these dependencies:

```xml

    <springfox-swagger2.version>2.9.2</springfox-swagger2.version>

[...]

    <dependency>
      <groupId>io.springfox</groupId>
      <artifactId>springfox-core</artifactId>
      <version>${springfox-swagger2.version}</version>
    </dependency>

    <dependency>
      <groupId>io.springfox</groupId>
      <artifactId>springfox-spi</artifactId>
      <version>${springfox-swagger2.version}</version>
    </dependency>

    <dependency>
      <groupId>io.springfox</groupId>
      <artifactId>springfox-spring-web</artifactId>
      <version>${springfox-swagger2.version}</version>
    </dependency>

    <dependency>
      <groupId>io.springfox</groupId>
      <artifactId>springfox-swagger2</artifactId>
      <version>${springfox-swagger2.version}</version>
    </dependency>

    <dependency>
      <groupId>io.springfox</groupId>
      <artifactId>springfox-data-rest</artifactId>
      <version>${springfox-swagger2.version}</version>
    </dependency>

    <dependency>
      <groupId>io.springfox</groupId>
      <artifactId>springfox-bean-validators</artifactId>
      <version>${springfox-swagger2.version}</version>
    </dependency>

    <dependency>
      <groupId>io.springfox</groupId>
      <artifactId>springfox-swagger-ui</artifactId>
      <version>${springfox-swagger2.version}</version>
    </dependency>
```

**Comment by ArchangelX360** [comment]
Indeed, I did the same but it should work if I read this correctly:  e8b0df7d31de4465d3638bd1e1e7b10a30794357

**Comment by kuwiano** [comment]
@ArchangelX360 Yep, definitely a bug. Btw. be careful this could break the schema generation (I ran into this issue with the `@ApiImplicitParam` annotation) - see #2204

**Comment by dilipkrish** [comment]
@ArchangelX360 you're right that it should work, but the only for spring data rest. For spring mvc, it isnt natively supported. Will add support in the upcoming release.

**Comment by tcolombo08** [comment]
The solution proposed by @falknerdominik [here](https://github.com/springfox/springfox/issues/2623#issuecomment-414297583) doesn't work for me. I'm using v2.9.2. Do you have an ETA on the next release @dilipkrish ?

**Comment by dilipkrish** [comment]
Mid November optimistically. I haven't had bandwidth recently to work on it, sorry.

**Comment by PaulHowes** [comment]
I just ran into this as well. Subscribing to the issue so I know when it's been fixed.

**Comment by stephaneeybert** [comment]
@MiguelAngelLV Lucky you. I tried the same this morning but it didn't help.

**Comment by stephaneeybert** [comment]
Here is the error I get:

> Caused by: org.springframework.beans.factory.support.BeanDefinitionOverrideException: Invalid bean definition with name 'pageableResolver' defined in class path resource [org/springframework/data/rest/webmvc/config/RepositoryRestMvcConfiguration.class]: Cannot register bean definition [Root bean: class [null]; scope=; abstract=false; lazyInit=false; autowireMode=3; dependencyCheck=0; autowireCandidate=true; primary=false; factoryBeanName=org.springframework.data.rest.webmvc.config.RepositoryRestMvcConfiguration; factoryMethodName=pageableResolver; initMethodName=null; destroyMethodName=(inferred); defined in class path resource [org/springframework/data/rest/webmvc/config/RepositoryRestMvcConfiguration.class]] for bean 'pageableResolver': There is already [Root bean: class [null]; scope=; abstract=false; lazyInit=false; autowireMode=3; dependencyCheck=0; autowireCandidate=true; primary=false; factoryBeanName=org.springframework.data.web.config.HateoasAwareSpringDataWebConfiguration; factoryMethodName=pageableResolver; initMethodName=null; destroyMethodName=(inferred); defined in class path resource [org/springframework/data/web/config/HateoasAwareSpringDataWebConfiguration.class]] bound.

**Comment by onacit** [comment]
@MiguelAngelLV Detailed explanation, please.

**Comment by hannes-angst** [comment]
Would be nice to have this feature.

What I came up a while ago was



    import io.swagger.annotations.ApiImplicitParam;
    import io.swagger.annotations.ApiImplicitParams;

    import java.lang.annotation.ElementType;
    import java.lang.annotation.Retention;
    import java.lang.annotation.RetentionPolicy;
    import java.lang.annotation.Target;

    @Target({ElementType.METHOD, ElementType.ANNOTATION_TYPE, ElementType.TYPE})
    @Retention(RetentionPolicy.RUNTIME)
    @ApiImplicitParams({
            @ApiImplicitParam(name = "page", dataType = "int", paramType = "query", defaultValue = "0", value = "Results page you want to retrieve (0..N)"),
            @ApiImplicitParam(name = "size", dataType = "int", paramType = "query", defaultValue = "20", value = "Number of records per page."),
            @ApiImplicitParam(name = "sort", allowMultiple = true, dataType = "string", paramType = "query", value = "Sorting criteria in the format: property(,asc|desc). "
                    + "Default sort order is ascending. " + "Multiple sort criteria are supported.")})
    public @interface ApiPageable {
    }



and in the controller:

    @ApiPageable
    public ResponseEntity<Page<MyObject>> listBots(@NonNull Pageable pageable) {
        return ResponseEntity.ok(service.myObject(pageable));
    }

**Comment by erdemedeiros** [comment]
As workaround if you're not using spring-data-rest plus ` @Import(SpringDataRestConfiguration.class)` you can register an `AlternateTypeRule` that's doing the same as [pageableConvension](https://github.com/springfox/springfox/blob/master/springfox-data-rest/src/main/java/springfox/documentation/spring/data/rest/configuration/SpringDataRestConfiguration.java#L49) in the swagger configuration:

```java
@Configuration
@EnableSwagger2
public class SwaggerConfig {

    private final TypeResolver typeResolver;
    private final RepositoryRestConfiguration restConfiguration;

    public SwaggerConfig(TypeResolver typeResolver, RepositoryRestConfiguration restConfiguration) {
        this.typeResolver = typeResolver;
        this.restConfiguration = restConfiguration;
    }

    @Bean
    @ConditionalOnMissingBean
    public Docket apiDocket() {
        return new Docket(DocumentationType.SWAGGER_2)
                .alternateTypeRules(newRule(typeResolver.resolve(Pageable.class),
                                            pageableMixin(restConfiguration),
                                            Ordered.HIGHEST_PRECEDENCE))
                ...
                .build();
    }

    private Type pageableMixin(RepositoryRestConfiguration restConfiguration) {
        return new AlternateTypeBuilder()
                .fullyQualifiedClassName(
                        String.format("%s.generated.%s",
                                      Pageable.class.getPackage().getName(),
                                      Pageable.class.getSimpleName()))
                .withProperties(Arrays.asList(
                        property(Integer.class,
                                 restConfiguration.getPageParamName()),
                        property(Integer.class,
                                 restConfiguration.getLimitParamName()),
                        property(String.class,
                                 restConfiguration.getSortParamName())
                ))
                .build();
    }

    private AlternateTypePropertyBuilder property(Class<?> type, String name) {
        return new AlternateTypePropertyBuilder()
                .withName(name)
                .withType(type)
                .withCanRead(true)
                .withCanWrite(true);
    }
```

**Comment by dkirrane** [comment]
Is there anyway to show only specific fields from Pageable? i.e. page, size and sort as string

At the moment I get every Pageable field in Swagger UI, even with the @ApiImplicitParams workarounds mentioned here. And their displayed in alphabetical order rather than the order in the code.

**Comment by franz-see** [comment]
I used @hannes-angst `@ApiPageable` and it works for me. I just have to add `@ApiIgnore` on my `Pageable` parameter though.

```
@ApiPageable
public ResponseEntity<Page<MyObject>> listBots(@ApiIgnore @NonNull Pageable pageable) {
    return ResponseEntity.ok(service.myObject(pageable));
}
```

**Comment by dilipkrish** [comment]
@franz-see thanks for the workaround solution!

**Comment by wwjay9** [comment]
My solution
```
@Bean
public Docket petApi() {
    return new Docket(DocumentationType.SWAGGER_2)
            .alternateTypeRules(AlternateTypeRules.newRule(Pageable.class, Paging.class))
            ...
}
```
```
@Getter
@Setter
@ApiModel("Pageing Query")
private static class Paging {

    @ApiModelProperty("pageNo default=0")
    private Integer page;

    @ApiModelProperty("PageSize default=10")
    private Integer size;
}
```

**Comment by sandeepsuvit** [comment]
> Would be nice to have this feature.
> 
> What I came up a while ago was
> 
> ```
> import io.swagger.annotations.ApiImplicitParam;
> import io.swagger.annotations.ApiImplicitParams;
> 
> import java.lang.annotation.ElementType;
> import java.lang.annotation.Retention;
> import java.lang.annotation.RetentionPolicy;
> import java.lang.annotation.Target;
> 
> @Target({ElementType.METHOD, ElementType.ANNOTATION_TYPE, ElementType.TYPE})
> @Retention(RetentionPolicy.RUNTIME)
> @ApiImplicitParams({
>         @ApiImplicitParam(name = "page", dataType = "int", paramType = "query", defaultValue = "0", value = "Results page you want to retrieve (0..N)"),
>         @ApiImplicitParam(name = "size", dataType = "int", paramType = "query", defaultValue = "20", value = "Number of records per page."),
>         @ApiImplicitParam(name = "sort", allowMultiple = true, dataType = "string", paramType = "query", value = "Sorting criteria in the format: property(,asc|desc). "
>                 + "Default sort order is ascending. " + "Multiple sort criteria are supported.")})
> public @interface ApiPageable {
> }
> ```
> 
> and in the controller:
> 
> ```
> @ApiPageable
> public ResponseEntity<Page<MyObject>> listBots(@NonNull Pageable pageable) {
>     return ResponseEntity.ok(service.myObject(pageable));
> }
> ```

For some strange reason i was getting a `io.swagger.models.parameters.AbstractSerializableParameter - Illegal DefaultValue  for parameter type integer` because of the `dataType="int"` and `defaultValue="0"` so i changed the dataType to `dataType="integer"` like below and the error doesn't show up:

```java
import io.swagger.annotations.ApiImplicitParam;
import io.swagger.annotations.ApiImplicitParams;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

@Target({ElementType.METHOD, ElementType.ANNOTATION_TYPE, ElementType.TYPE})
@Retention(RetentionPolicy.RUNTIME)
@ApiImplicitParams({
        @ApiImplicitParam(name = "page", dataType = "integer", paramType = "query", defaultValue = "0", value = "Results page you want to retrieve (0..N)"),
        @ApiImplicitParam(name = "size", dataType = "integer", paramType = "query", defaultValue = "20", value = "Number of records per page."),
        @ApiImplicitParam(name = "sort", allowMultiple = true, dataType = "string", paramType = "query", value = "Sorting criteria in the format: property(,asc|desc). "
                + "Default sort order is ascending. " + "Multiple sort criteria are supported.")})
public @interface ApiPageable {
}
```
Not sure if this is a bug or something

**Comment by steklopod** [comment]
You don't need to place annotation in each function:
> Kotlin:

```kotlin
data class SwaggerPageable(
    @ApiModelProperty("Number of records per page", example = "20")
    val size: Int?,
    @ApiModelProperty("Results page you want to retrieve (0..N)", example = "0")
    val page: Int?,
    @ApiModelProperty(
        "Sorting criteria in the format: property(,asc|desc). Default sort order is ascending. Multiple sort criteria are supported.",
        example = "&sort=created,asc"
    )
    var sort: String?
)

    @Bean
    fun api(): Docket = Docket(DocumentationType.SWAGGER_2)
        .groupName("Cards")
        .apiInfo(getApiInfo())
        .select()
        .apis(RequestHandlerSelectors.basePackage(controllerFolder))
        .paths(PathSelectors.any())
        .build()
        .directModelSubstitute(Pageable::class.java, SwaggerPageable::class.java)
```

**Comment by dilipkrish** [comment]
Since there is plenty of suggested work arounds, Im pushing this to the next patch since I dont want to cram this important feature just before release

**Comment by dirkluijk** [comment]
Interesting note: the `name` and `defaultValue` are not the same for every Spring project, they are dynamic because they can be configured.

e.g.

```yaml
spring:
  data:
    web:
      pageable:
        page-parameter: page
        size-parameter: page-size
        one-indexed-parameters: true
        default-page-size: 10
        max-page-size: 50
```

**Comment by naineshmp** [comment]
After adding @ApiPageable like this, 

```java
@Target({ElementType.METHOD, ElementType.ANNOTATION_TYPE, ElementType.TYPE})
@Retention(RetentionPolicy.RUNTIME)
@ApiImplicitParams({
  @ApiImplicitParam(
      name = "page_size",
      dataType = "int",
      paramType = "query",
      value =
          "OPTIONAL The preferred number of workflow runs to return in a page. If not provided, the implementation should use a default page size. The implementation must not return more items than `page_size`, but it may return fewer.  Clients should not assume that if fewer than `page_size` items are returned that all items have been returned.  The availability of additional pages is indicated by the value of `next_page_token` in the response."),
  @ApiImplicitParam(
      name = "page_token",
      dataType = "int",
      paramType = "query",
      value =
          "OPTIONAL Token to use to indicate where to start getting results. If unspecified, return the first page of results."),
  @ApiImplicitParam(
      name = "sort",
      allowMultiple = true,
      dataType = "string",
      paramType = "query",
      value =
          "OPTIONAL Sorting criteria in the format: property(,asc|desc). ")
})
public @interface ApiPageable {}
```

generated swagger-json file is missing **type** which is breaking the Swagger-Gen automation down the line. 

![image](https://user-images.githubusercontent.com/26369197/110514107-cf4b7280-80bb-11eb-8eae-2294ded48176.png)

Not sure what I am configuring wrong - which is resulting in "no-type". Can anyone help?

**Comment by naineshmp** [comment]
> After adding @ApiPageable like this,
> 
> ```java
> @Target({ElementType.METHOD, ElementType.ANNOTATION_TYPE, ElementType.TYPE})
> @Retention(RetentionPolicy.RUNTIME)
> @ApiImplicitParams({
>   @ApiImplicitParam(
>       name = "page_size",
>       dataType = "int",
>       paramType = "query",
>       value =
>           "OPTIONAL The preferred number of workflow runs to return in a page. If not provided, the implementation should use a default page size. The implementation must not return more items than `page_size`, but it may return fewer.  Clients should not assume that if fewer than `page_size` items are returned that all items have been returned.  The availability of additional pages is indicated by the value of `next_page_token` in the response."),
>   @ApiImplicitParam(
>       name = "page_token",
>       dataType = "int",
>       paramType = "query",
>       value =
>           "OPTIONAL Token to use to indicate where to start getting results. If unspecified, return the first page of results."),
>   @ApiImplicitParam(
>       name = "sort",
>       allowMultiple = true,
>       dataType = "string",
>       paramType = "query",
>       value =
>           "OPTIONAL Sorting criteria in the format: property(,asc|desc). ")
> })
> public @interface ApiPageable {}
> ```
> 
> generated swagger-json file is missing **type** which is breaking the Swagger-Gen automation down the line.
> 
> ![image](https://user-images.githubusercontent.com/26369197/110514107-cf4b7280-80bb-11eb-8eae-2294ded48176.png)
> 
> Not sure what I am configuring wrong - which is resulting in "no-type". Can anyone help?

After replacing 'dataType' with 'dataTypeClass' , It works fine. 

```
@Target({ElementType.METHOD, ElementType.ANNOTATION_TYPE, ElementType.TYPE})
@Retention(RetentionPolicy.RUNTIME)
@ApiImplicitParams({
  @ApiImplicitParam(
      name = "page_size",
      dataTypeClass = Integer.class,
      paramType = "query",
      value =
          "OPTIONAL The preferred number of workflow runs to return in a page. If not provided, the implementation should use a default page size. The implementation must not return more items than `page_size`, but it may return fewer.  Clients should not assume that if fewer than `page_size` items are returned that all items have been returned.  The availability of additional pages is indicated by the value of `next_page_token` in the response."),
  @ApiImplicitParam(
      name = "page_token",
      dataTypeClass = Integer.class,
      paramType = "query",
      value =
          "OPTIONAL Token to use to indicate where to start getting results. If unspecified, return the first page of results."),
  @ApiImplicitParam(
      name = "sort",
      allowMultiple = true,
      dataTypeClass = String.class,
      paramType = "query",
      value =
          "OPTIONAL Sorting criteria in the format: property(,asc|desc). ")
})
public @interface ApiPageable {}
```

**Comment by souravsharma1210** [comment]
None of the work around is working for me.

ERROR s.d.s.w.s.ApiDescriptionReader : Skipping process path[/tasks/search], method[search] as it has an error.
java.lang.NoSuchMethodError: springfox.documentation.spi.service.contexts.ParameterContext.<init>(Lspringfox/documentation/service/ResolvedMethodParameter;Lspringfox/documentation/builders/ParameterBuilder;Lspringfox/documentation/spi/service/contexts/DocumentationContext;Lspringfox/documentation/spi/schema/GenericTypeNamingStrategy;Lspringfox/documentation/spi/service/contexts/OperationContext

Pom Swagger:

                <dependency>
		<groupId>io.springfox</groupId>
		<artifactId>springfox-boot-starter</artifactId>
		<version>3.0.0</version>
		 </dependency>
		 @PostMapping("/search1")
	
	@ApiOperation(value = "Search task")
	public ResponseEntity<Page<Resource>> search1(
			@RequestParam UUID unit,
			@RequestParam UUID id,
			@RequestParam Type type, @RequestParam String query,Pageable page) {
		
		return ResponseEntity.ok(null);
	}
		The above api is skipped in swagger-ui .if I remove the Pageable in parameter list. It works fine.
		

@dilipkrish Can u please help here.

**Comment by philbannon** [comment]
Hi,

Is there any update on this issue? Still seeing the original issue 

Thanks

**Comment by lArtiquel** [comment]
No updates on this bug?

**Comment by Duduxs** [comment]
Any update? :)

**Comment by minditer1** [comment]
Don't know if this helps anybody, but I fixed this issue by bumping to 1.6.5 version of the OpenAPI doc and with this code:

````
@Parameter(hidden = true)
@ParameterObject
@PageableDefault(sort = {DEFAULT_SORT_COLUMN}, direction = Sort.Direction.DESC) Pageable pageable
````

**Comment by Of1mjkee** [comment]
> As workaround, if you're not using spring-data-rest plus ` @Import(SpringDataRestConfiguration.class)`, you can register an `AlternateTypeRule` that's doing the same as [pageableConvension](https://github.com/springfox/springfox/blob/master/springfox-data-rest/src/main/java/springfox/documentation/spring/data/rest/configuration/SpringDataRestConfiguration.java#L49) in the swagger configuration:
> 
> ```java
> @Configuration
> @EnableSwagger2
> public class SwaggerConfig {
> 
>     private final TypeResolver typeResolver;
>     private final RepositoryRestConfiguration restConfiguration;
> 
>     public SwaggerConfig(TypeResolver typeResolver, RepositoryRestConfiguration restConfiguration) {
>         this.typeResolver = typeResolver;
>         this.restConfiguration = restConfiguration;
>     }
> 
>     @Bean
>     @ConditionalOnMissingBean
>     public Docket apiDocket() {
>         return new Docket(DocumentationType.SWAGGER_2)
>                 .alternateTypeRules(newRule(typeResolver.resolve(Pageable.class),
>                                             pageableMixin(restConfiguration),
>                                             Ordered.HIGHEST_PRECEDENCE))
>                 ...
>                 .build();
>     }
> 
>     private Type pageableMixin(RepositoryRestConfiguration restConfiguration) {
>         return new AlternateTypeBuilder()
>                 .fullyQualifiedClassName(
>                         String.format("%s.generated.%s",
>                                       Pageable.class.getPackage().getName(),
>                                       Pageable.class.getSimpleName()))
>                 .withProperties(Arrays.asList(
>                         property(Integer.class,
>                                  restConfiguration.getPageParamName()),
>                         property(Integer.class,
>                                  restConfiguration.getLimitParamName()),
>                         property(String.class,
>                                  restConfiguration.getSortParamName())
>                 ))
>                 .build();
>     }
> 
>     private AlternateTypePropertyBuilder property(Class<?> type, String name) {
>         return new AlternateTypePropertyBuilder()
>                 .withName(name)
>                 .withType(type)
>                 .withCanRead(true)
>                 .withCanWrite(true);
>     }
> ```

@erdemedeiros, Thanks!  spring-data-web + springfox-boot-starter 3.0.0

```java
    @Bean
    public AlternateTypeRuleConvention pageableConvention(
            final TypeResolver resolver,
            final SpringDataWebProperties properties) {
        return new AlternateTypeRuleConvention() {

            @Override
            public int getOrder() {
                return Ordered.HIGHEST_PRECEDENCE;
            }

            @Override
            public List<AlternateTypeRule> rules() {
                return List.of(
                        newRule(resolver.resolve(Pageable.class), resolver.resolve(pageableMixin(properties)))
                );
            }
        };
    }

    private Type pageableMixin(final SpringDataWebProperties properties) {
        return new AlternateTypeBuilder()
                .fullyQualifiedClassName(
                        String.format("%s.generated.%s",
                                Pageable.class.getPackage().getName(),
                                Pageable.class.getSimpleName()))
                .property(b -> property(b, Integer.class, properties.getPageable().getPageParameter()))
                .property(b -> property(b, Integer.class, properties.getPageable().getSizeParameter()))
                .property(b -> property(b, String.class, properties.getSort().getSortParameter()))
                .build();
    }

    private AlternateTypePropertyBuilder property(AlternateTypePropertyBuilder builder, Class<?> type, String name) {
        return builder
                .name(name)
                .type(type)
                .canRead(true)
                .canWrite(true);
    }
```

**Comment by CrissCS** [comment]
Hi all,

Has there been any official fix for this issue?

**Comment by dorianpavetic** [comment]
> Don't know if this helps anybody, but I fixed this issue by bumping to 1.6.5 version of the OpenAPI doc and with this code:
> 
> ```
> @Parameter(hidden = true)
> @ParameterObject
> @PageableDefault(sort = {DEFAULT_SORT_COLUMN}, direction = Sort.Direction.DESC) Pageable pageable
> ```

Why does this comment has no upvotes? This `@ParameterObject` annotation is the right solution for me (`@Parameter(hidden = true)` not even needed though) - spring boot 3.2.4 and OpenAPI 3.1.0
